pinot-broker/src/main/java/org/apache/pinot/broker/routing/instanceselector/MultiStageStrictReplicaGroupSelector.java

```diff
@@ -0,0 +1,150 @@
+import org.apache.pinot.common.metrics.BrokerMetrics;
+import org.apache.pinot.spi.config.table.TableType;
+import org.apache.pinot.spi.config.table.assignment.InstancePartitionsType;
+import org.apache.pinot.spi.utils.builder.TableNameBuilder;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * Instance selector for multi-stage queries which can ensure that Colocated Tables always leverage Colocated Join
+ * whenever possible. To achieve this, this instance-selector uses InstancePartitions (IP) to determine replica-groups,
+ * as opposed to IdealState used by other instance-selectors. Moreover, this also uses the requestId generated by
+ * Pinot broker to determine the replica-group picked for each table involved in the query, as opposed to using a
+ * member variable. There may be scenarios where an instance in the chosen replica-group is down. In that case, this
+ * strategy will try to pick another replica-group. For realtime tables, this strategy uses only CONSUMING partitions.
+ * This is feature is in <strong>Beta</strong>.
+ */
+public class MultiStageStrictReplicaGroupSelector extends BaseInstanceSelector {
+  private static final Logger LOGGER = LoggerFactory.getLogger(MultiStageStrictReplicaGroupSelector.class);
+
+  private final String _tableNameWithType;
+  private final ZkHelixPropertyStore<ZNRecord> _propertyStore;
+  private InstancePartitions _instancePartitions;
+
+  public MultiStageStrictReplicaGroupSelector(String tableNameWithType, ZkHelixPropertyStore<ZNRecord> propertyStore,
+      BrokerMetrics brokerMetrics, @Nullable AdaptiveServerSelector adaptiveServerSelector) {
+    super(tableNameWithType, brokerMetrics, adaptiveServerSelector);
+    _tableNameWithType = tableNameWithType;
+    _propertyStore = propertyStore;
+  }
+
+  @Override
+  public void init(Set<String> enabledInstances, IdealState idealState, ExternalView externalView,
+      Set<String> onlineSegments) {
+    super.init(enabledInstances, idealState, externalView, onlineSegments);
+    _instancePartitions = getInstancePartitions();
+  }
+
+  @Override
+  public void onInstancesChange(Set<String> enabledInstances, List<String> changedInstances) {
+    super.onInstancesChange(enabledInstances, changedInstances);
+    _instancePartitions = getInstancePartitions();
+  }
+
+  @Override
+  public void onAssignmentChange(IdealState idealState, ExternalView externalView, Set<String> onlineSegments) {
+    super.onAssignmentChange(idealState, externalView, onlineSegments);
+    _instancePartitions = getInstancePartitions();
+  }
+
+  @Override
+  Map<String, String> select(List<String> segments, int requestId,
+      Map<String, List<String>> segmentToEnabledInstancesMap, Map<String, String> queryOptions) {
+    // Create a copy of InstancePartitions to avoid race-condition with event-listeners above.
+    InstancePartitions localInstancePartitions = _instancePartitions.createCopy();
```

Comment:
We don't need to create a copy here. Instead, you may define `_instancePartitions` as `volatile`, and assign it to a local variable
```suggestion
    InstancePartitions instancePartitions = _instancePartitions;
```

Reply:
Yeah good point. I think in a previous commit I was modifying these but then I changed to a simple reference switch. Fixed.